plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3ClientSettingsTests.java

```diff
@@ -0,0 +1,129 @@
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Map;
+
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.isEmptyString;
+import static org.hamcrest.Matchers.nullValue;
+
+public class S3ClientSettingsTests extends ESTestCase {
+    public void testThereIsADefaultClientByDefault() {
+        final Map<String, S3ClientSettings> settings = S3ClientSettings.load(Settings.EMPTY);
+        assertThat(settings.keySet(), contains("default"));
+
+        final S3ClientSettings defaultSettings = settings.get("default");
+        assertThat(defaultSettings.credentials, nullValue());
+        assertThat(defaultSettings.endpoint, isEmptyString());
+        assertThat(defaultSettings.protocol, is(Protocol.HTTPS));
+        assertThat(defaultSettings.proxyHost, isEmptyString());
+        assertThat(defaultSettings.proxyPort, is(80));
+        assertThat(defaultSettings.proxyUsername, isEmptyString());
+        assertThat(defaultSettings.proxyPassword, isEmptyString());
+        assertThat(defaultSettings.readTimeoutMillis, is(ClientConfiguration.DEFAULT_SOCKET_TIMEOUT));
+        assertThat(defaultSettings.maxRetries, is(ClientConfiguration.DEFAULT_RETRY_POLICY.getMaxErrorRetry()));
+        assertThat(defaultSettings.throttleRetries, is(ClientConfiguration.DEFAULT_THROTTLE_RETRIES));
+    }
+
+    public void testDefaultClientSettingsCanBeSet() {
+        final Map<String, S3ClientSettings> settings = S3ClientSettings.load(Settings.builder()
+            .put("s3.client.default.max_retries", 10).build());
+        assertThat(settings.keySet(), contains("default"));
+
+        final S3ClientSettings defaultSettings = settings.get("default");
+        assertThat(defaultSettings.maxRetries, is(10));
+    }
+
+    public void testNondefaultClientCreatedBySettingItsSettings() {
+        final Map<String, S3ClientSettings> settings = S3ClientSettings.load(Settings.builder()
+            .put("s3.client.another_client.max_retries", 10).build());
+        assertThat(settings.keySet(), contains("default", "another_client"));
+
+        final S3ClientSettings defaultSettings = settings.get("default");
+        assertThat(defaultSettings.maxRetries, is(ClientConfiguration.DEFAULT_RETRY_POLICY.getMaxErrorRetry()));
+
+        final S3ClientSettings anotherClientSettings = settings.get("another_client");
+        assertThat(anotherClientSettings.maxRetries, is(10));
+    }
+
+    public void testRejectionOfLoneAccessKey() {
+        final IllegalArgumentException e = expectThrows(IllegalArgumentException.class,
+            () -> {
+                final MockSecureSettings secureSettings = new MockSecureSettings();
+                secureSettings.setString("s3.client.default.access_key", "aws_key");
+                S3ClientSettings.load(Settings.builder().setSecureSettings(secureSettings).build());
```

Comment:
Nit: you can build the settings outside the expectThrows block, I find it easier to "spot" where the exception is expected to be thrown

Reply:
++ I pushed ee41de633714c0797dd2823595d4a46afa6c68b6